test/e2e/clusterctl_upgrade.go

```diff
@@ -0,0 +1,300 @@
+		// So we are creating a workload cluster that will be used as a new management cluster where to install older version of providers
+
+		clusterctl.ApplyClusterTemplateAndWait(ctx, clusterctl.ApplyClusterTemplateAndWaitInput{
+			ClusterProxy: input.BootstrapClusterProxy,
+			ConfigCluster: clusterctl.ConfigClusterInput{
+				LogFolder:                filepath.Join(input.ArtifactFolder, "clusters", input.BootstrapClusterProxy.GetName()),
+				ClusterctlConfigPath:     input.ClusterctlConfigPath,
+				KubeconfigPath:           input.BootstrapClusterProxy.GetKubeconfigPath(),
+				InfrastructureProvider:   clusterctl.DefaultInfrastructureProvider,
+				Flavor:                   clusterctl.DefaultFlavor,
+				Namespace:                managementClusterNamespace.Name,
+				ClusterName:              fmt.Sprintf("%s-%s", specName, util.RandomString(6)),
+				KubernetesVersion:        input.E2EConfig.GetVariable(KubernetesVersion),
+				ControlPlaneMachineCount: pointer.Int64Ptr(1),
+				WorkerMachineCount:       pointer.Int64Ptr(1),
+			},
+			WaitForClusterIntervals:      input.E2EConfig.GetIntervals(specName, "wait-cluster"),
+			WaitForControlPlaneIntervals: input.E2EConfig.GetIntervals(specName, "wait-control-plane"),
+			WaitForMachineDeployments:    input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
+		}, managementClusterResources)
+
+		By("Turning the workload cluster into a management cluster with older versions of providers")
+
+		// In case of the cluster id a DockerCluster, we should load controller images into the nodes.
+		// Nb. this can be achieved also by changing the DockerMachine spec, but for the time being we are using
+		// this approach because this allows to have a single source of truth for images, the e2e config
+		// Nb. the images for official version of the providers will be pulled from internet, but the latest images must be
+		// built locally and loaded into kind
+		cluster := managementClusterResources.Cluster
+		if cluster.Spec.InfrastructureRef.Kind == "DockerCluster" {
+			Expect(bootstrap.LoadImagesToKindCluster(ctx, bootstrap.LoadImagesToKindClusterInput{
+				Name:   cluster.Name,
+				Images: input.E2EConfig.Images,
+			})).To(Succeed())
+		}
+
+		// Get a ClusterBroker so we can interact with the workload cluster
+		managementClusterProxy = input.BootstrapClusterProxy.GetWorkloadCluster(ctx, cluster.Namespace, cluster.Name)
+
+		// Download the v1alpha3 clusterctl version to be used for setting up the management cluster to be upgraded
+		clusterctlBinaryURL := input.E2EConfig.GetVariable(initWithBinaryVariableName)
+		clusterctlBinaryURL = strings.ReplaceAll(clusterctlBinaryURL, "{OS}", runtime.GOOS)
+		clusterctlBinaryURL = strings.ReplaceAll(clusterctlBinaryURL, "{ARCH}", runtime.GOARCH)
+
+		log.Logf("downloading clusterctl binary from %s", clusterctlBinaryURL)
+		binaryFileclusterctlBinaryPath := downloadToTmpFile(clusterctlBinaryURL)
```

Comment:
```suggestion
		clusterctlBinaryPath := downloadToTmpFile(clusterctlBinaryURL)
```
Maybe that would be better